File: ts/src/delegation.ts
/**
 * ROAR Protocol — Capability delegation tokens.
 *
 * Mirrors python/src/roar_sdk/delegation.py exactly.
 * Uses Node.js 18+ built-in crypto — no external dependencies.
 *
 * Signing format: Ed25519 over canonical JSON of token fields
 * (same sort-keys approach as HMAC message signing).
 */

import {
  createPrivateKey,
  createPublicKey,
  sign as cryptoSign,
  verify as cryptoVerify,
  randomBytes,
} from "crypto";

// ---------------------------------------------------------------------------
// Canonical JSON (sort keys, mirrors Python json.dumps(sort_keys=True))
// ---------------------------------------------------------------------------

function sortedJson(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return "null";
    if (Number.isInteger(value)) return `${value}.0`;
    return String(value);
  }
  if (Array.isArray(value)) {
    if (value.length === 0) return "[]";
    return "[" + value.map(sortedJson).join(", ") + "]";
  }
  if (typeof value === "object") {
    const keys = Object.keys(value as Record<string, unknown>).sort();
    if (keys.length === 0) return "{}";
    const pairs = keys.map(
      (k) => `${JSON.stringify(k)}: ${sortedJson((value as Record<string, unknown>)[k])}`
    );
    return "{" + pairs.join(", ") + "}";
  }
  return String(value);
}

// ---------------------------------------------------------------------------
// DER key helpers (same as signing.ts)
// ---------------------------------------------------------------------------

function privateKeyFromHex(hex: string): ReturnType<typeof createPrivateKey> {
  const raw = Buffer.from(hex, "hex");
  const pkcs8Header = Buffer.from("302e020100300506032b657004220420", "hex");
  return createPrivateKey({
    key: Buffer.concat([pkcs8Header, raw]),
    format: "der",
    type: "pkcs8",
  });
}

function publicKeyFromHex(hex: string): ReturnType<typeof createPublicKey> {
  const raw = Buffer.from(hex, "hex");
  const spkiHeader = Buffer.from("302a300506032b6570032100", "hex");
  return createPublicKey({
    key: Buffer.concat([spkiHeader, raw]),
    format: "der",
    type: "spki",
  });
}

// ---------------------------------------------------------------------------
// DelegationToken
// ---------------------------------------------------------------------------

export interface DelegationToken {
  /** "tok_" + 10 random hex chars */
  token_id: string;
  delegator_did: string;
  delegate_did: string;
  capabilities: string[];
  issued_at: number;
  expires_at: number | null;
  max_uses: number | null;
  use_count: number;
  can_redelegate: boolean;
  signature: string;
}

/** Return true if the token has not expired and has remaining uses. */
export function isTokenValid(token: DelegationToken): boolean {
  const now = Date.now() / 1000;
  if (token.expires_at !== null && now > token.expires_at) return false;
  if (token.max_uses !== null && token.use_count >= token.max_uses) return false;
  return true;
}

/** Return true if the token grants the requested capability. */
export function tokenGrants(token: DelegationToken, capability: string): boolean {
  return token.capabilities.includes(capability) || token.capabilities.includes("*");
}

/**
 * Atomically check validity and consume one use of the token (increment use_count).
 * Returns true if use was recorded, false if the token is exhausted.
 *
 * This is a single synchronous check+increment with no await between them,
 * making it safe against TOCTOU in JS's single-threaded event loop.
 * Do NOT call isTokenValid() separately before this — use verifyAndValidateToken()
 * to combine signature + validity + consume in one logical step.
 */
export function consumeToken(token: DelegationToken): boolean {
  if (token.max_uses !== null && token.use_count >= token.max_uses) return false;
  token.use_count += 1;
  return true;
}

// ---------------------------------------------------------------------------
// Signing body — all fields except `signature` and `use_count`
// ---------------------------------------------------------------------------

function signingBody(token: DelegationToken): Buffer {
  const body: Record<string, unknown> = {
    token_id: token.token_id,
    delegator_did: token.delegator_did,
    delegate_did: token.delegate_did,
    capabilities: token.capabilities,
    issued_at: token.issued_at,
    expires_at: token.expires_at,
    max_uses: token.max_uses,
    can_redelegate: token.can_redelegate,
  };
  return Buffer.from(sortedJson(body), "utf-8");
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Issue a new DelegationToken signed by the delegator's Ed25519 private key.
 *
 * @param delegatorDid        - DID of the issuing agent
 * @param delegatorPrivateKey - 32-byte raw private key as hex
 * @param delegateDid         - DID of the recipient agent
 * @param capabilities        - list of capability strings (e.g. ["read", "write"])
 * @param expiresInSeconds    - TTL from now (default 3600). 0 = no expiry.
 * @param maxUses             - optional hard use cap
 * @param canRedelegate       - whether the delegate may further delegate
 */
export function issueToken(
  delegatorDid: string,
  delegatorPrivateKey: string,
  delegateDid: string,
  capabilities: string[],
  expiresInSeconds = 3600,
  maxUses: number | null = null,
  canRedelegate = false,
): DelegationToken {
  const now = Date.now() / 1000;
  const token: DelegationToken = {
    token_id: "tok_" + randomBytes(5).toString("hex"),
    delegator_did: delegatorDid,
    delegate_did: delegateDid,
    capabilities,
    issued_at: now,
    expires_at: expiresInSeconds > 0 ? now + expiresInSeconds : null,
    max_uses: maxUses,
    use_count: 0,
    can_redelegate: canRedelegate,
    signature: "",
  };

  const privKey = privateKeyFromHex(delegatorPrivateKey);
  const rawSig = cryptoSign(null, signingBody(token), privKey) as Buffer;
  token.signature = "ed25519:" + rawSig.toString("base64url");
  return token;
}

/**
 * Verify a DelegationToken's Ed25519 signature only.
 *
 * @param token               - the token to verify
 * @param delegatorPublicKey  - 32-byte raw public key as hex (must match token.delegator_did's key)
 * @returns true if signature is cryptographically valid
 * @note Does NOT check expiry or use count — use verifyAndValidateToken for full validation.
 */
export function verifyToken(token: DelegationToken, delegatorPublicKey: string): boolean {
  if (!token.signature.startsWith("ed25519:")) return false;
  try {
    const pubKey = publicKeyFromHex(delegatorPublicKey);
    const b64 = token.signature.slice("ed25519:".length);
    const rawSig = Buffer.from(b64, "base64url");
    return cryptoVerify(null, signingBody(token), pubKey, rawSig) as boolean;
  } catch {
    return false;
  }
}

/**
 * Verify signature AND check that the token is not expired or exhausted.
 * Prefer this over calling verifyToken + isTokenValid separately (avoids TOCTOU).
 *
 * @returns true only if signature valid AND not expired AND uses remaining
 */
export function verifyAndValidateToken(token: DelegationToken, delegatorPublicKey: string): boolean {
  return verifyToken(token, delegatorPublicKey) && isTokenValid(token);
}

File: python/src/roar_sdk/delegation.py
# -*- coding: utf-8 -*-
"""ROAR Protocol — Delegation tokens for scoped capability grants (Layer 1).

A delegation token allows an agent (the delegator) to grant a subset of
their capabilities to another agent (the delegate) for a limited time or
number of uses. The token is signed with the delegator's Ed25519 private key,
so the delegate can prove the grant to third parties without the delegator
being online.

Usage::

    from roar_sdk.signing import generate_keypair
    from roar_sdk.delegation import DelegationToken, issue_token, verify_token

    # Delegator issues a token
    priv, pub = generate_keypair()
    token = issue_token(
        delegator_did="did:roar:agent:alice-...",
        delegator_private_key=priv,
        delegate_did="did:roar:agent:bob-...",
        capabilities=["recon", "scan"],
        expires_in_seconds=3600,
        max_uses=10,
    )

    # Delegate includes the token in their messages:
    msg.context["delegation_token"] = token.model_dump()

    # Third party verifies:
    ok = verify_token(token, delegator_public_key=pub)
"""

from __future__ import annotations

import base64
import json
import time
import uuid
from typing import List, Optional

from pydantic import BaseModel, Field


class DelegationToken(BaseModel):
    """A signed capability grant from one agent to another."""

    token_id: str = Field(default_factory=lambda: f"tok_{uuid.uuid4().hex[:10]}")
    delegator_did: str
    delegate_did: str
    capabilities: List[str]  # subset of delegator's capabilities
    issued_at: float = Field(default_factory=time.time)
    expires_at: Optional[float] = None   # None = no expiry
    max_uses: Optional[int] = None       # None = unlimited
    use_count: int = 0
    can_redelegate: bool = False         # can the delegate further delegate?
    signature: str = ""                  # "ed25519:<base64url>"

    def is_valid(self) -> bool:
        """Check expiry and use count (does NOT verify signature)."""
        if self.expires_at is not None and time.time() > self.expires_at:
            return False
        if self.max_uses is not None and self.use_count >= self.max_uses:
            return False
        return True

    def grants(self, capability: str) -> bool:
        """Check if this token grants a specific capability."""
        return self.is_valid() and capability in self.capabilities

    def consume(self) -> bool:
        """Record one use of this token. Returns False if the token is exhausted.

        Call this every time the token is accepted so max_uses is enforced.
        """
        if not self.is_valid():
            return False
        self.use_count += 1
        return True

    def _signing_body(self) -> bytes:
        """Canonical body for signing — deterministic, sorted keys."""
        body = json.dumps(
            {
                "capabilities": sorted(self.capabilities),
                "can_redelegate": self.can_redelegate,
                "delegate_did": self.delegate_did,
                "delegator_did": self.delegator_did,
                "expires_at": self.expires_at,
                "issued_at": self.issued_at,
                "max_uses": self.max_uses,
                "token_id": self.token_id,
            },
            sort_keys=True,
        )
        return body.encode("utf-8")


def issue_token(
    delegator_did: str,
    delegator_private_key: str,
    delegate_did: str,
    capabilities: List[str],
    *,
    expires_in_seconds: Optional[float] = 3600.0,
    max_uses: Optional[int] = None,
    can_redelegate: bool = False,
) -> DelegationToken:
    """Issue and sign a delegation token.

    Args:
        delegator_did: The issuing agent's DID.
        delegator_private_key: 64-char hex Ed25519 private key.
        delegate_did: The receiving agent's DID.
        capabilities: List of capability strings being granted.
        expires_in_seconds: TTL from now (None = no expiry).
        max_uses: Maximum number of uses (None = unlimited).
        can_redelegate: Whether the delegate can further delegate.

    Returns:
        A signed DelegationToken.

    Raises:
        ImportError: If cryptography package is not installed.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
    except ImportError:
        raise ImportError(
            "Delegation tokens require the 'cryptography' package. "
            "Install it: pip install 'roar-sdk[ed25519]'"
        )

    expires_at = (time.time() + expires_in_seconds) if expires_in_seconds else None

    token = DelegationToken(
        delegator_did=delegator_did,
        delegate_did=delegate_did,
        capabilities=capabilities,
        expires_at=expires_at,
        max_uses=max_uses,
        can_redelegate=can_redelegate,
    )

    private = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(delegator_private_key))
    raw_sig = private.sign(token._signing_body())
    sig_b64 = base64.urlsafe_b64encode(raw_sig).decode("ascii").rstrip("=")
    token.signature = f"ed25519:{sig_b64}"
    return token


def verify_token(
    token: DelegationToken,
    delegator_public_key: str,
) -> bool:
    """Verify a delegation token's signature and validity.

    Args:
        token: The token to verify.
        delegator_public_key: 64-char hex Ed25519 public key of the delegator.

    Returns:
        True if the signature is valid and the token has not expired or been exhausted.
    """
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
    except ImportError:
        raise ImportError(
            "Delegation tokens require the 'cryptography' package. "
            "Install it: pip install 'roar-sdk[ed25519]'"
        )

    if not token.is_valid():
        return False

    sig_value = token.signature
    if not sig_value.startswith("ed25519:"):
        return False

    b64 = sig_value[len("ed25519:"):]
    padding = (4 - len(b64) % 4) % 4
    raw_sig = base64.urlsafe_b64decode(b64 + "=" * padding)

    try:
        public = Ed25519PublicKey.from_public_bytes(bytes.fromhex(delegator_public_key))
        public.verify(raw_sig, token._signing_body())
        return True
    except (InvalidSignature, ValueError):
        return False
